js/modules/loki_message_api.js

```diff
@@ -155,6 +177,90 @@ class LokiMessageAPI {
     log.info(`Successful storage message to ${pubKey}`);
   }
 
+  async *retrieveNextMessage(nodeUrl) {
+    const params = {
+      pubKey: ourKey,
+      lastHash: nodeData.lastHash || '',
+    };
+    const options = {
+      timeout: 40000,
+      headers: {
+        [LOKI_LONGPOLL_HEADER]: true,
+      },
+    };
+    while (true) {
+      const result = await rpc(
+        `http://${nodeUrl}`,
+        this.snodeServerPort,
+        'retrieve',
+        params,
+        options
+      );
+      if (Array.isArray(result.messages) && result.messages.length) {
+        const filteredMessages = await this.jobQueue.add(() =>
+          filterIncomingMessages(result.messages)
+        );
+        if (filteredMessages.length) {
+          yield filteredMessages;
+        }
+      }
+    }
+  }
+
+  async openConnection(callback) {
+    while (this.ourSwarmNodes.length > 0) {
+      const url = this.ourSwarmNodes.pop();
+      const successive_failures = 0;
+      while (true) {
+        // loop breaks upon error
+        try {
+          for await (let messages of retrieveNextMessages(url)) {
```

Comment:
Maybe we could move 
```js
lokiSnodeAPI.updateLastHash(
  url,
  lastMessage.hash,
  lastMessage.expiration
);
callback(messages);
successive_failures = 0;
```
inside `retrieveNextMessages` and it wouldn't need to be a generator?

Reply:
You're right that there are too many loops and that the generator might be encapsulating too much. I'll try to simplify things.

Reply:
Changed